rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+    state_store: Arc<state_store::Client<C>>,
+    lease_name: Vec<u8>,
+    lease_holder_name: Vec<u8>,
+    current_fencing_token: Arc<Mutex<Option<HybridLogicalClock>>>,
+    auto_renewal_cancellation_token: CancellationToken,
+}
+
+impl<C> Drop for Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync,
+    C::PubReceiver: Send + Sync,
+{
+    fn drop(&mut self) {
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+    }
+}
+
+/// Key Lease client implementation
+///
+/// Notes:
+/// Do not call any of the methods of this client after the `state_store` parameter is shutdown.
+/// Calling any of the methods in this implementation after the `state_store` is shutdown results in undefined behavior.
+impl<C> Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync,
+    C::PubReceiver: Send + Sync,
+{
+    /// Create a new Key Lease Client.
+    ///
+    /// Notes:
+    /// - `lease_holder_name` is expected to be the client ID used in the underlying MQTT connection settings.
+    /// - There must be one instance of `leased_lock::Client` per lease.
+    ///
+    /// # Errors
+    /// [`struct@Error`] of kind [`KeyLengthZero`](ErrorKind::KeyLengthZero) if the `lease_name` is empty
+    ///
+    /// [`struct@Error`] of kind [`LeaseHolderNameLengthZero`](ErrorKind::LeaseHolderNameLengthZero) if the `lease_holder_name` is empty
+    pub fn new(
+        state_store: Arc<state_store::Client<C>>,
+        lease_name: Vec<u8>,
+        lease_holder_name: Vec<u8>,
+    ) -> Result<Self, Error> {
+        if lease_name.is_empty() {
+            return Err(Error(ErrorKind::KeyLengthZero));
+        }
+
+        if lease_holder_name.is_empty() {
+            return Err(Error(ErrorKind::LeaseHolderNameLengthZero));
+        }
+
+        Ok(Self {
+            state_store,
+            lease_name,
+            lease_holder_name,
+            current_fencing_token: Arc::new(Mutex::new(None)),
+            auto_renewal_cancellation_token: CancellationToken::new(),
+        })
+    }
+
+    /// Gets the latest fencing token related to the most recent lease.
+    ///
+    /// Returns either None or an actual Fencing Token (`HybridLogicalClock`).
+    /// None means that either a lease has not been acquired previously with this client, or
+    /// if a lease renewal has failed (if lease auto-renewal is used). The presence of a `HybridLogicalClock`
+    /// does not mean that it is the most recent (and thus valid) Fencing Token - in case
+    /// auto-renewal has not been used and the lease has already expired.
+    ///
+    /// # Panics
+    /// Possible panic if `std::Lock::lock()` fails.
+    #[must_use]
+    pub fn get_current_lease_fencing_token(&self) -> Option<HybridLogicalClock> {
+        self.current_fencing_token
+            .lock()
+            .expect("Could not lock mutex")
+            .clone()
+    }
+
+    async fn internal_acquire(
+        &mut self,
```

Comment:
mut needed? (If not, same comment for the calling functions)

Reply:
Interesting. I though because a field is being set, `mut` would be needed. I need to reconcile that.

Changed.

Reply:
it's because the field is under an Arc<Mutex> - technically you aren't modifying the Arc, you're just changing what it references to